src/java_bytecode/java_static_initializers.cpp

```diff
@@ -8,11 +8,38 @@ Author: Chris Smowton, chris.smowton@diffblue.com
 
 #include "java_static_initializers.h"
 #include "java_object_factory.h"
+#include "java_types.h"
 #include <goto-programs/class_hierarchy.h>
 #include <util/std_types.h>
 #include <util/std_expr.h>
 #include <util/std_code.h>
 #include <util/suffix.h>
+#include <util/arith_tools.h>
+
+/// The three states in which a `<clinit>` method for a class can be before,
+/// after, and during static class initialization.
+///
+/// According to the JVM Spec document (section 5.5), the JVM needs to
+/// maintain, for every class initializer, a state indicating whether the
+/// initializer has been executed, is being executed, or has raised errors.
+/// The spec mandates that the JVM consider 4 different states (not
+/// initialized, being initialized, ready for use, and initialization error).
+/// The `clinit_statet` is a simplification of those 4 states where:
+///
+/// `NOT_INIT` corresponds to "not initialized"
+/// `IN_PROGRESS` corresponds to "some thread is currently running the
+///   `<clinit>` and no other thread should run it"
+/// `RETURN` corresponds to "the `<clinit>` has been executed and the class
+///   is ready to be used, or it has errored"
+///
+/// The last state corresponds to a fusion of the two original states "ready
+/// for use" and "initialization error".
+enum class clinit_statest
+{
+  NOT_INIT,
+  IN_PROGRESS,
+  RETURN
```

Comment:
INIT_COMPLETE?

Reply:
instead of 'RETURN' yes INIT_COMPLETE is better, changing.

Reply:
@Degiorgio Do not forget to update the documentation above